cache/models/subject.js

```diff
@@ -0,0 +1,228 @@
+/**
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * cache/model/subject.js
+ */
+'use strict'; // eslint-disable-line strict
+
+const sampleStore = require('../sampleStore');
+const constants = sampleStore.constants;
+const redisClient = require('../redisCache').client.sampleStore;
+const u = require('../../utils/filters');
+const ONE = 1;
+const TWO = 2;
+
+// using the bluebird promise implementation instead of the native promise
+const Promise = require('bluebird');
+
+/*
+ * All the query params that can be expected in the hierarchy endpoint are
+ * defined as a key. The values of the query parameters are later added to them
+ * by the set filters function
+ */
+const filters = {
+  aspect: {},
+  subjectTags: {},
+  aspectTags: {},
+  status: {},
+};
+
+/**
+ * Get the aspect information from the redis and attach it to the sample object
+ * @param  {String} aspectName - Aspect name
+ * @returns {Promise} - which resolves to an aspect object
+ */
+function getAspectFromRedis(aspectName) {
+  const aspectKey = sampleStore.toKey(constants.objectType.aspect,
+                              aspectName);
+  return redisClient.hgetallAsync(aspectKey)
+    .then((aspect) => {
+      const fieldsToStringify = constants.fieldsToStringify.aspect;
+      if (aspect) {
+        fieldsToStringify.forEach((field) => {
+          if (aspect[field] && !Array.isArray(aspect[field])) {
+            aspect[field] = JSON.parse(aspect[field]);
+          }
+        });
+      }
+
+      return aspect;
+    });
+} // getAspectFromRedis
+
+/**
+ * Get the sample information from Redis
+ * @param  {String} sampleName - Sample name
+ * @returns {Promise} - which resolves to a sample object
+ */
+function getSampleFromRedis(sampleName) {
+  const sampleKey = sampleStore.toKey(constants.objectType.sample,
+                              sampleName);
+  return redisClient.hgetallAsync(sampleKey)
+    .then((samp) => {
+      const fieldsToStringify = constants.fieldsToStringify.sample;
+      if (samp) {
+        fieldsToStringify.forEach((field) => {
+          if (samp[field] && !Array.isArray(samp[field])) {
+            samp[field] = JSON.parse(samp[field]);
+          }
+        });
+      }
+
+      return samp;
+    });
+} // getSampleFromReids
+
+/**
+ * Given a subject with its samples, aspect, aspectTags and sampleStatus filters
+ * are applied to the samples and the filtered samples are attached back to the
+ * subject
+ * @param  {Object} subj - Subject object on which the filters are applied
+ * @returns {Object} - filtered subject
+ */
+function filterSamples(subj) {
+  const filteredSamples = [];
```

Comment:
@iamigo @pallavi2209  I moved the aspect filters along with the other filters. Applying just the aspect includes filter while getting it from redis (using "hmget aspectnames")  will work but adding the excludes filter will nest the promises and break the promise chain. Any thoughts about this?